### DISCUSSION TITLE
Agent Custom Fields don't appear to be populating when included in a URL Action

### DISCUSSION THREAD

**Discussion by Zebble** [amidaware/tacticalrmm#1230] [General]
**Server Info (please complete the following information):**
 - OS: Debian 11.4
 - Browser: Chrome, FireFox
 - RMM Version: 0.14.3

**Installation Method:**
  - [X] Standard
  - [ ] Docker

**Agent Info (please complete the following information):**
- Agent version: 2.1.1
- Agent OS: Windows Server 2019

**Describe the bug**

We've created an Agent custom field in TRMM called "ScreenConnectGUID".
We've confirmed this is being properly populated on all agents, using the community-script called "Screenconnect - Get GUID for client". 
We then have a URL Action that links to our ScreenConnect instance:  https://screenconnect.instance.domain/########//{{agent.ScreenConnectGUID}}/Join
This used to work, but stopped working in at least 0.14.2.  It leaves {{agent.ScreenConnectGUID}} blank.  We've confirmed it's passing other non-custom agent values ({{agent.hostname}}, for instance) just fine by using a bogus URL for testing.

**Comment by wh1te909** [comment]
I am not able to reproduce, please watch video below and let me know if this is what you're trying to do:

https://user-images.githubusercontent.com/7434746/181365525-99e7fbc9-281b-48df-92b0-5499f9594fd7.mp4

**Comment by Zebble** [comment]
Hi @wh1te909!

Thanks for making that video.

That's exactly what we have in place and the ScreenconnectGUID custom field is being properly populated.  It just doesn't show up in the URL Action when added.  This was working very well at least a couple of weeks ago for us.

The only changes we made were to upgrade to 0.14.2 then 0.14.3 and we also migrated from a Docker environment to a native Debian environment.  That was fun.  We had a few issues but seem to have worked through them and everything else seems to be working as expected.

**Comment by dinger1986** [comment]
out of interest how did you migrate from docker to native?

**Comment by Zebble** [comment]
Good question @dinger1986.  I didn't think I'd have to do it again, so I didn't document, but from memory:

1) Installed a fresh native version in a clean Debian environment.
1a)  Made sure that the Docker and Debian environments were running the same (latest) version.
2) Looked at the docker-compose.yaml for "volumes" and started figuring out where this data was in the native version.  Compared our running docker environment data folders to the native folders. (most stuff in Debian sits in the /rmm folder)
3) Shut down Docker on the current environment and stopped all services on the Debian environment (Mesh, Tactical, Redis, PostgreSQL, etc.).
3a)  Postgres is a newer version in the Debian install (14 vs 13) so you may need to do an upgrade of the DB in this.  I had already been using 14 in the Docker version so didn't have to mess around with this.
4) Copied (rsync) the data from the Docker environment found in step 2 to the new environment.  I'd suggest backing up/renaming the original folders in Debian before you overwrite them.  At the very least, keep a copy of the Debian version of /rmm/api/tacticalrmm/tacticalrmm/local_settings.py
5) Edit /rmm/api/tacticalrmm/tacticalrmm/local_settings.py so it matches the simpler native version format of Tactical.  Mine ended up being:

```
SECRET_KEY = '<redacted>'

DEBUG = False

ALLOWED_HOSTS = ['api.example.com', 'tactical-backend']

ADMIN_URL = '<redacted>'

CORS_ORIGIN_WHITELIST = [
    'https://rmm.example.com'
]

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql',
        'NAME': 'tacticalrmm',
        'USER': 'redacted',
        'PASSWORD': 'redacted',
        'HOST': 'localhost',
        'PORT': '5432',
    }
}

MESH_USERNAME = 'tactical'
MESH_SITE = 'https://mesh.example.com'
REDIS_HOST    = 'localhost'
ADMIN_ENABLED = False

MON_TOKEN = "<redacted>"

MESH_TOKEN_KEY = "<redacted>"
```

The key in my case was there was a "DOCKER_BUILD = True" that was hanging around from the original local_settings.py that caused issues.  There were some other fields as well from the docker build that were not necessary.  The Debian-based settings files is much simpler and I kept it as clean as I could, with any original data from the Docker config.


6)  Force a tactical update.  Not sure if this is 100% needed, but I did it anyway.  (./update.sh --force)

7)  Services should be started as part of the update, so things should start working.

8)  Change your DNS entries and test!

One issue I had was Mesh not connecting.  This was due to the extra Docker stuff in the settings file.  The troubleshooting tips at https://docs.tacticalrmm.com/troubleshooting/ were very helpful.

**Comment by Zebble** [comment]
@dinger1986, if it helps, this is the list of volumes/locations I had to address in the docker install:

/var/lib/docker/volumes/rmm_postgres_data/_data
/var/lib/docker/volumes/rmm_redis_data/_data
/var/lib/docker/volumes/rmm_tactical_data/_data
/var/lib/docker/volumes/rmm_mesh_data/_data
/var/lib/docker/volumes/rmm_mongo_data/_data

Then had to locate where these were in the Debian install by comparing their contents.

**Comment by dinger1986** [comment]
Nice thanks! Appreciated

**Comment by Zebble** [comment]
No worries!  You're welcome.  If you do run into problems, happy to assist.

**Comment by wh1te909** [comment]
@Zebble you did pretty well on the migration lol it was actually a lot simpler than that but not your fault obv cuz we don't have an official guide. there are only 3 files needed to migrate, `local_settings.py` (removing the docker stuff like you did) and then the dump files generated by `mongodump` and `pg_dump` which btw a dump from postgres 13 can be imported into postgres 14 that takes care of updating the db.
 
Im not sure if you still have access to the original instance if so then you should try to dump the databases of both and use that to migrate instead of trying to copy data from volumes. It sounds like something didn't migrate properly which is why you're having this bug

**Reply by Zebble** [reply]
Thanks @wh1te909.  If I ever need to migrate again, i'll do that.  :)  Unfortunately, I did the migration over a week ago and we've since added a bunch more stuff to TRMM so I can't really re-migrate.  I'll start going through the variable-replacement code for the URL Action and see where my setup might be missing something.  Any advice on where best to start in the code?

**Comment by Zebble** [comment]
Ok, fixed the issue.  I thought I had done this prior, but I guess not.
- I removed the "ScreenConnectGUID" Agent custom field.
- Added a new "ScreenConnectGUID" Agent custom field.
- Re-associated the new field to the script in the "Tasks" for the respective policy.
- Ran the task manually on a few agents.
- Confirmed the GUID is being captured into the new custom field.
- Tried the original URL Action.
- Presto, it works again!

Not sure what was actually wrong, but starting from scratch fixed it.
